torch_geometric/nn/models/nbfnet.py

```diff
@@ -0,0 +1,191 @@
+from functools import reduce
+
+import torch
+from torch import LongTensor, Tensor
+from torch.nn import Embedding, LayerNorm, Module, ModuleList, ReLU
+
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.nn.dense.linear import Linear
+from torch_geometric.nn.models import MLP
+from torch_geometric.typing import Optional, OptTensor, Tuple
+from torch_geometric.utils import add_self_loops
+
+
+def edge_match(edge_index, query_index):
+    # source of algorithm:
+    # https://github.com/KiddoZhu/NBFNet-PyG/blob/master/nbfnet/tasks.py#L6
+    # we prepare a unique hashing of edges, mapping them to long ints
+    # idea: max number of edges = num_nodes * num_relations
+    # given a tuple (h, r), we search for all other existing edges
+    # starting from head h with relation r
+    base = edge_index.max(dim=1)[0] + 1
+    # we make sure that the maximum product is less than MAX_LONG_INT
+    assert reduce(int.__mul__, base.tolist()) < torch.iinfo(torch.long).max
+    scale = base.cumprod(0)
+    scale = torch.div(scale[-1], scale, rounding_mode='floor')
+    # map both the original edge index and the query index to long ints
+    edge_hash = (edge_index * scale.unsqueeze(-1)).sum(dim=0)
+    edge_hash, order = edge_hash.sort()
+    query_hash = (query_index * scale.unsqueeze(-1)).sum(dim=0)
+    # matched ranges: [start[i], end[i])
+    start = torch.bucketize(query_hash, edge_hash)
+    end = torch.bucketize(query_hash, edge_hash, right=True)
+    # num_match shows how many edges satisfy the (h, r) pattern
+    # for each query in the batch
+    num_match = end - start
+    # generate the corresponding ranges
+    offset = num_match.cumsum(0) - num_match
+    range = torch.arange(num_match.sum(), device=edge_index.device)
+    range = range + (start - offset).repeat_interleave(num_match)
+    return order[range], num_match
+
+
+def index_to_mask(index, size):
+    index = index.view(-1)
+    size = int(index.max()) + 1 if size is None else size
+    mask = index.new_zeros(size, dtype=torch.bool)
+    mask[index] = True
+    return mask
+
+
+class NBFConv(MessagePassing):
+    def __init__(self, in_channels: int, out_channels: int, **kwargs):
+        kwargs.setdefault('aggr', 'add')
+        kwargs.setdefault('node_dim', 0)
+        super().__init__(**kwargs)
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.relation = Embedding(1, in_channels)
+        self.linear = Linear(2 * in_channels, out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.linear.reset_parameters()
+        self.relation.reset_parameters()
+
+    def forward(self, x: Tensor, boundary: Tensor, edge_index: LongTensor,
+                edge_weight: OptTensor = None) -> Tensor:
+        x0 = x.clone()
+        x = self.propagate(edge_index, x=x, edge_weight=edge_weight)
+        x = x + boundary
+        x = torch.cat([x0, x], dim=-1)
+        x = self.linear(x)
```

Comment:
```suggestion
        out = self.propagate(edge_index, x=x, edge_weight=edge_weight)
        out = out + boundary
        out = torch.cat([out, x], dim=-1)
        out = self.linear(out)
```

Reply:
Sorry, I've been quite busy lately. I'm planning to add the tests and docs in late September.
(If you ask whether you're allowed to add the test and docs instead of me, yes absolutely!)